wordpress-mobile/WordPress-Android
# QR Code Auth: Part 7 - Implement FluxC Validate and Authenticate calls

Parent #16481 

This PR adds the following:
- Hooks in the FluxC call to QRCodeAuthStore.Validate call. This returns the location/browser that will be shown in the validated view.
- Hooks in the FluxC call to QRCodeAuthStore.Authenticate call
Note: The FluxC PR is https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/2392
- Adds tests into QRCodeAuthViewModelTest 

Notes: 
- This PR will be merged to a feature branch and not trunk. Milestone will be set to "future"

**Merge Instructions**
- Ensure that #16711 has been merged to the feature branch
- Merge as normal

Prior to testing this PR you will need to
- Make sure your **real device** is sandboxed 
- Make sure your sandbox is pointed to D79766-code (apply the patch)
- Have access to the "Calypso live (direct link)" in the related Calypso PR: 
    - https://github.com/Automattic/wp-calypso/pull/63257
    

**Tests** (These tests were lifted from the iOS [PR](https://github.com/wordpress-mobile/WordPress-iOS/pull/18756) and modified where needed for Android)

**Pre-req**
- Install the app
- Login using a WP.com account
- Navigate to Me -> App Settings -> Debug Settings
- Enable `QRCodeAuthFlowFeatureConfig`
- Restart the App

Test A: Scanning a token with invalid data
![invalid-qrlogin-code](https://user-images.githubusercontent.com/793774/170778949-f8b47a11-ce27-4db0-87f5-85f5be09ec07.png)
1. Launch the app
2. Navigate to My Site > Me > Scan Login Code
3. Scan the QR code image from above
4. ✅ Verify you see the "Could not validate" error view
5. ✅ Verify tapping Scan again takes you to the scanning view again

Test B: Scanning a valid but expired token
![qrlogin-expired](https://user-images.githubusercontent.com/793774/170781480-d8fc1f46-6052-4e67-94e7-257b96689562.png)
1. Launch the app
2. Navigate to My Site > Me > Scan Login Code
3. Scan the QR code image from above
4. ✅  Verify you see the "Expired log in code" error view
5. ✅ Verify tapping Scan again takes you to the scanning view again

Test C: Scanning a valid token
1. Launch the app and login with a wp.com account
2. Navigate to My Site > Me > Scan Login Code
3. Scan the QR Code from the "Calypso live (direct link)" in the Calypso PR (See above). **Make sure you are not logged in to WP.com when you go to the direct link. If you use an incognito window to open the link, the you can't verify step 12.
4. ✅  Verify you see a "Are you trying to login" message that displays your web browser and approximate location
5. "Are you trying to login to Chrome near Los Angeles, California?"
6. ✅  Verify tapping Cancel exits out of the flow and nothing else happens
7. Scan the valid code again
8. Tap the 'Yes, log me in' button
9. ✅  Verify you see a loading indicator with the current view faded out
10. ✅  Verify after a second you see the "You're logged in!" message
11. ✅  Verify tapping Dismiss exits the flow
12. ✅  Verify you are logged in on your web browser

Test D: Testing no internet error
1. Launch the app
3. Navigate to My Site > Me > Scan Login Code
4. Turn on Airplane mode
5. Scan the QR Code from the related Calypso PR (See above)
6. ✅ Verify you see a "No connection" error state
7. ✅ Tapping Scan again brings you back to the scanning view
8. Enable your internet and scan again
9. ✅  Verify the scanning works correctly 


## Regression Notes
1. Potential unintended areas of impact
N/A

2. What I did to test those areas of impact (or what existing automated tests I relied on)
Manual and unit testing

3. What automated tests I added (or what prevented me from doing so)
QRCodeAuthViewModelTest

PR submission checklist:
